app/src/test/java/org/dhis2/usescases/teiDashboard/dashboardfragments/relationships/RelationshipPresenterTest.kt

```diff
@@ -0,0 +1,485 @@
+                .byTrackedEntityInstance().eq("tei_to_uid")
+                .byTrackedEntityAttribute().`in`(listOf("tea_uid"))
+                .blockingGet()
+        ) doReturn trackedEntityAttributeValue()
+
+        presenter.updateRelationships()
+
+        verify(view).setRelationships(relationshipViewModelsTo())
+    }
+
+    @Test
+    fun `Should set relationship types in function of teiType`() {
+        val relationshipTypes =
+            listOf(
+                Trio.create(
+                    RelationshipType.builder()
+                        .uid("relationship_type").build(),
+                    "tei_type", 0
+                )
+            )
+
+        whenever(
+            dashboardRepository
+                .relationshipsForTeiType("tei_type")
+        ) doReturn relationshipForTeiType()
+
+        whenever(view.abstracContext) doReturn mock()
+        whenever(
+            dashboardRepository
+                .getObjectStyle(view.abstracContext, "tei_type")
+        ) doReturn 0
+
+        presenter.relationshipForTeiType()
+
+        verify(view).setRelationshipTypes(relationshipTypes)
+    }
+
+    @Test
+    fun `Should delete a relationship`() {
+        whenever(
+            d2.relationshipModule().relationships().withItems()
+                .uid("relationship_uid")
+        ) doReturn mock()
+        whenever(view.analyticsHelper()) doReturn mock()
+
+        val testSubscriber = presenter.updateRelationships.test()
+        presenter.deleteRelationship(Relationship.builder().uid("relationship_uid").build())
+
+        testSubscriber.assertValueCount(1)
+        testSubscriber.assertValue(true)
+    }
+
+    @Test
+    fun `Should add a relationship`() {
+        val testingRelationshipUid = "note_uid"
+        whenever(
+            d2.relationshipModule().relationships().blockingAdd(
+                RelationshipHelper.teiToTeiRelationship("tei_uid", "tei_to", "type")
+            )
+        ) doReturn testingRelationshipUid
+
+        val testSubscriber = presenter.updateRelationships.test()
+        presenter.addRelationship("tei_to", "type")
+
+        testSubscriber.assertValueCount(1)
+        testSubscriber.assertValue(true)
+    }
+
+    @Test
+    fun `Should call to go to add relationship`() {
+        whenever(
+            d2.programModule().programs().uid("program_uid")
+                .blockingGet()
+        ) doReturn getProgramDefaultAccessTrue()
+        whenever(view.analyticsHelper()) doReturn mock()
+
+        presenter.goToAddRelationship("type_add")
+
+        verify(view).goToAddRelationship("tei_uid", "type_add")
+    }
+
+    @Test
+    fun `Should open dashboard`() {
+        whenever(
+            d2.trackedEntityModule().trackedEntityInstances()
+                .byUid().eq("tei_uid")
+        ) doReturn mock()
+        whenever(
+            d2.trackedEntityModule().trackedEntityInstances()
+                .byUid().eq("tei_uid").one()
+        ) doReturn mock()
```

Comment:
Is this necessary when the mock of `d2` has the `RETURN_DEEP_STUBS` configuration?

Reply:
I had trouble with the "eq" because RETURN_DEEP_STUBS  does not mocked and I needed to do in two parts

Reply:
great!